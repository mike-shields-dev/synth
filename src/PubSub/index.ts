export { ControlChangeSubscriber, publishControlChange } from './ControlChange';
export type { ControlChange } from './ControlChange/types';
export { MidiNoteOnSubscriber, publishMidiNoteOn } from './MidiNoteOn';
export type { MidiNoteOn } from './MidiNoteOn/types';
export { NoteOffSubscriber, publishNoteOff } from './NoteOff';
export type { NoteOff } from './NoteOff/types';

