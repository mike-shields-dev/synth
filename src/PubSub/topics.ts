const
    MIDI = 'midi',
    MIDI_CC = 'midi.controlChange',
    MIDI_NOTE_ON = 'midi.note.on',
    MIDI_NOTE_OFF = 'midi.note.off',
    UI_CC = 'ui.controlChange',
    FOCUS = 'focus';

export {
    MIDI, 
    MIDI_CC, 
    MIDI_NOTE_ON,
    MIDI_NOTE_OFF,
    UI_CC,
    FOCUS,
}