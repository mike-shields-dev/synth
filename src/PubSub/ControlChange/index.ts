import PubSub from 'pubsub-js';
import { CONTROL } from '../topics';
import { ControlChange, ControlChangeHandler } from './types';

function publishControlChange(data: ControlChange) {
  PubSub.publish(CONTROL, data);
}

class ControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: ControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      CONTROL,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishControlChange,
  ControlChangeSubscriber
};

