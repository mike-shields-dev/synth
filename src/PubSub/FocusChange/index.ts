import PubSub from "pubsub-js";
import { FOCUS } from '../topics';
import { FocusChangeHandler } from "./types";

function publishFocusChange(data: string) {
    PubSub.publish(FOCUS, data);
}

class FocusChangeSubscriber {
    private subscription: string;

    constructor(focusHandler: FocusChangeHandler) {
        this.subscription = PubSub.subscribe(
            FOCUS,
            focusHandler,
        );
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    }
}

export {
    publishFocusChange,
    FocusChangeSubscriber
};
