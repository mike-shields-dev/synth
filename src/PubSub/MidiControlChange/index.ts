import PubSub from 'pubsub-js';
import { MIDI_CC } from '../topics';
import { MidiControlChange, MidiControlChangeHandler } from './types';

function publishMidiControlChange(data: MidiControlChange) {
  PubSub.publish(MIDI_CC, data);
}

class MidiControlChangeSubscriber {
  private subscription: string;

  constructor(controlChangeHandler: MidiControlChangeHandler) {
    this.subscription = PubSub.subscribe(
      MIDI_CC,
      controlChangeHandler,
    );
  }

  public unsubscribe() {
    PubSub.unsubscribe(this.subscription);
  }
}

export {
  publishMidiControlChange,
  MidiControlChangeSubscriber
};

