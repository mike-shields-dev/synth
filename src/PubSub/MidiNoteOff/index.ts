import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { MidiNoteOff, MidiNoteOffHandler } from './types';

function publishMidiNoteOff(payload: MidiNoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, payload)
}

class MidiNoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: MidiNoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOff, MidiNoteOffSubscriber };
