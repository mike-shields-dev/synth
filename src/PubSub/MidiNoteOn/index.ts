import PubSub from "pubsub-js";
import { MIDI_NOTE_ON } from "../topics";
import { MidiNoteOn, MidiNoteOnHandler } from "./types";

function publishMidiNoteOn(payload: MidiNoteOn) {
    PubSub.publish(MIDI_NOTE_ON, payload)
}

class MidiNoteOnSubscriber {
    private subscription: any;

    constructor(noteOnHandler: MidiNoteOnHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_ON,
            noteOnHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishMidiNoteOn, MidiNoteOnSubscriber };
