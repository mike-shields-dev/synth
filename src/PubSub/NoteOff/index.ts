import PubSub from "pubsub-js";
import { MIDI_NOTE_OFF } from "../topics";
import { NoteOff, NoteOffHandler } from './types';

function publishNoteOff(data: NoteOff) {
    PubSub.publish(MIDI_NOTE_OFF, data)
}

class NoteOffSubscriber {
    private subscription: string;

    constructor(noteOffHandler: NoteOffHandler) {
        this.subscription = PubSub.subscribe(
            MIDI_NOTE_OFF,
            noteOffHandler,
        )
    }

    public unsubscribe() {
        PubSub.unsubscribe(this.subscription);
    };
}

export { publishNoteOff, NoteOffSubscriber };
