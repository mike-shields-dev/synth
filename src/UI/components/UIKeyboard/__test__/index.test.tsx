import { UIKeyboard } from "..";
import { fireEvent, render, screen, waitFor, act} from '@testing-library/react';
import { MIDI_NOTE_OFF, MIDI_NOTE_ON } from "../../../../PubSub/topics";
import { vi } from 'vitest';


describe('UIKeyboard', () => {
    beforeEach(() => {
        render(<UIKeyboard />);
    });

    it('matches snapshot', () => {
        const { container } = render(<UIKeyboard />)
        
        expect(container).toMatchSnapshot();
    });

    it('renders 12 buttons, one for each piano key of an octave', () => {
        expect(screen.getAllByRole('button')).toHaveLength(12);
    });

    it('each key publishes a note on message with the correct noteNumber', async () => {
        const spy = vi.fn();
        const keys = screen.getAllByRole('button', { name: 'key' });
        const subscriber = PubSub.subscribe(MIDI_NOTE_ON, spy);

        for(const key of keys) {
            const keyValue = Number(key.getAttribute('value'));
            await act(() => fireEvent.mouseDown(key))

            await waitFor(() => {
                expect(spy).toHaveBeenCalledWith(MIDI_NOTE_ON, {
                    noteNumber: keyValue,
                    velocity: 80,
                });
            });
        };

        PubSub.unsubscribe(subscriber);
    });

    it('each key publishes a note off message with the correct note number', async () => {
        const spy = vi.fn();
        const keys = screen.getAllByRole('button', { name: 'key' });
        const subscriber = PubSub.subscribe(MIDI_NOTE_OFF, spy);

        for(const key of keys) {
            const keyValue = Number(key.getAttribute('value'));
            await act(() => fireEvent.mouseUp(key))

            await waitFor(() => {
                expect(spy).toHaveBeenCalledWith(MIDI_NOTE_OFF, {
                    noteNumber: keyValue,
                });
            });
        };

        PubSub.unsubscribe(subscriber);
    });
});