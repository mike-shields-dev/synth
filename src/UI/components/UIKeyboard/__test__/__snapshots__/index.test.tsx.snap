// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UIKeyboard > matches snapshot 1`] = `
<div>
  <div>
    <label
      for="transpose"
    >
      Octave: 
    </label>
    <input
      id="transpose"
      max="5"
      min="-5"
      type="number"
      value="0"
    />
    <div>
      <button
        aria-label="key"
        type="button"
        value="60"
      >
        C
      </button>
      <button
        aria-label="key"
        type="button"
        value="61"
      >
        Db
      </button>
      <button
        aria-label="key"
        type="button"
        value="62"
      >
        D
      </button>
      <button
        aria-label="key"
        type="button"
        value="63"
      >
        Eb
      </button>
      <button
        aria-label="key"
        type="button"
        value="64"
      >
        E
      </button>
      <button
        aria-label="key"
        type="button"
        value="65"
      >
        F
      </button>
      <button
        aria-label="key"
        type="button"
        value="66"
      >
        Gb
      </button>
      <button
        aria-label="key"
        type="button"
        value="67"
      >
        G
      </button>
      <button
        aria-label="key"
        type="button"
        value="68"
      >
        Ab
      </button>
      <button
        aria-label="key"
        type="button"
        value="69"
      >
        A
      </button>
      <button
        aria-label="key"
        type="button"
        value="70"
      >
        Bb
      </button>
      <button
        aria-label="key"
        type="button"
        value="71"
      >
        B
      </button>
    </div>
  </div>
</div>
`;
